### DISCUSSION TITLE
what is the correct way to use mongoose-autopopulate on query objects.

### DISCUSSION THREAD

**Discussion by lostrouter** [typegoose/typegoose#593] [Q&A]
I'm trying to leverage `mongoose-autopopulate` in my project and i'm running into an issue with the autopopulated reference passing through the right types.   Typescript sees the field as a `Ref<ObjectId>`.  i can get around this by some interesting typecasting, or abuse of json.parse/json.stringify.  What is the correct way to use `mongoose-autopopulate` in this instance so that i get the fields of the ref passed through for typescript to use them?

```ts
@plugin(autopopulate)
export class SubscriptionPlan extends Base {

  @prop({ required: true })
  public name!: string;

}

export class Company extends Base {

  @prop({ required: true })
  public companyName!: string

  @prop({ autopopulate: true, ref: () => SubscriptionPlan })
  public companySubscriptionPlan?: Ref<SubscriptionPlan>;

}

const searchedCompany = await CompanyModel.findOne({
    _id: new ObjectID(companyId),
  });


companyPlanName = company.companySubscriptionPlan
        ? JSON.parse(JSON.stringify(company.companySubscriptionPlan)).name
        : ''
```

intellisense
```
(property) Company.companySubscriptionPlan?: Ref<SubscriptionPlan, Types.ObjectId>
```

ts error
```
Property 'name' does not exist on type 'Ref<SubscriptionPlan, ObjectId>'.
  Property 'name' does not exist on type 'ObjectId'.
```

## Versions

- System: macos 
- NodeJS: 14.0.0
- Typescript: 3.9.2

- Compiler / Transpiler: tsc / ts-node / ts-jest 
- Typegoose(NPM): 7.6.0 
- mongoose: 5.10.18 
- mongodb: 3.3.0-beta2

---

**Comment by hasezoey** [comment]
typegoose sets the type as `Ref` because it can be either type and typescript provides something called "typeguards", which typegoose also has for this purpose:

- [`isDocument`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document)
- [`isRef`](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-ref-type)

(Note: they are not interchangeable (`!isDocument != isRef`))

**Reply by lostrouter** [reply]
if i understand you correctly. i should be able to do this?

```
companyPlanName = isDocument(company.companySubscriptionPlan) ?
        company.companySubscriptionPlan.name
        : ''
```

**Reply by hasezoey** [reply]
yes that should be *a* correct way, you could also use the following:

```ts
if (isDocument(doc.reference)) {
  // in this block "doc.reference" should be the correct document type
}
// not here anymore

// OR

// import "assertion" from typegoose or define your own with correct types
assertion(isDocument(doc.reference), new Error("Your Custom Error Here")); // use this in tests or if you use the throw-error way to early-return (like in rust)
// in this block (the same "assertion" is in), "doc.reference" should be the correct document type
```
